### PR TITLE
Adding more logging to StoreStats and CompactionPolicy

### PR DESCRIPTION
**Pull request by nsivabalan**
SLA: 10 mins
Reviewers: Gopal

### CODE REVIEW

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 1]
```diff
@@ -355,6 +357,13 @@ public void close() {
       }
       metrics.statsOnDemandScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTimeMs,
           TimeUnit.MILLISECONDS);
+      indexSegmentCount++;
+      if (indexSegmentCount == 1) {
+        logger.info("Completed scanning of 1st index segment by on-demand scanning {} " + indexSegment);
+      }
+      if (indexSegmentCount % 10 == 0) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: Think we should just collapse this to 
```
if (indexSegmentCount == 1 || indexSegmentCount % 10 == 0) {
  // print info (maybe name of the last processed segment)
}
```

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 2]
```diff
@@ -366,9 +375,11 @@ public void close() {
    * @return a {@link NavigableMap} of log segment name to valid data size
    */
   private NavigableMap<String, Long> collectValidDataSizeByLogSegment(long referenceTimeInMs) throws StoreException {
+    logger.trace("On demand index scanning to collect compaction data stats wrt ref time " + referenceTimeInMs);
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: formatting

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 3]
```diff
@@ -382,6 +393,13 @@ public void close() {
       }
       metrics.statsOnDemandScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTimeMs,
           TimeUnit.MILLISECONDS);
+      indexSegmentCount++;
+      if (indexSegmentCount == 1) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: same here

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 4]
```diff
@@ -747,13 +767,16 @@ public void run() {
...
         Offset firstCheckpoint = index.getCurrentEndOffset();
+        logger.trace("First checkpoint by IndexScanner {} ", firstCheckpoint);
         ConcurrentNavigableMap<Offset, IndexSegment> indexSegments = index.getIndexSegments();
         Map<StoreKey, Long> deletedKeys = new HashMap<>();
         // process the active index segment based on the firstCheckpoint in case index segment rolled over after the
         // checkpoint is taken
         if (!cancelled && indexSegments.size() > 0) {
           Map.Entry<Offset, IndexSegment> activeIndexSegmentEntry = indexSegments.floorEntry(firstCheckpoint);
           long indexSegmentStartProcessTime = time.milliseconds();
+          logger.trace("Fetching active index entries {} in active segments before first checkpoint ",
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: this log message doesn't look very clear.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 5]
```diff
@@ -771,13 +796,24 @@ public void run() {
               processIndexSegmentEntriesBackward(indexSegment, indexEntries, deletedKeys);
               metrics.statsBucketingScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTime,
                   TimeUnit.MILLISECONDS);
+              segmentCount++;
+              if (segmentCount == 1) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: collapse

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 6]
```diff
@@ -867,6 +904,7 @@ private void forwardScan(Offset startOffset, Offset endOffset) throws StoreExcep
      */
     private void processIndexSegmentEntriesBackward(IndexSegment indexSegment, List<IndexEntry> indexEntries,
         Map<StoreKey, Long> deletedKeys) throws StoreException {
+      logger.trace("Processing index entries backward by IndexScanner for segment " + indexSegment);
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: formatting

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 7]
```diff
@@ -358,11 +358,8 @@ public void close() {
       metrics.statsOnDemandScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTimeMs,
           TimeUnit.MILLISECONDS);
       indexSegmentCount++;
-      if (indexSegmentCount == 1) {
-        logger.info("Completed scanning of 1st index segment by on-demand scanning {} " + indexSegment);
-      }
-      if (indexSegmentCount % 10 == 0) {
-        logger.info("Completed scanning batch of 10 index segments by on-demand scanning {} " + indexSegment);
+      if (indexSegmentCount == 1 || indexSegmentCount % 10 == 0) {
+        logger.info("Index segment processing(on-demand scanning) for container stats complete for {} " + indexSegment);
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: remove the "+"

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 8]
```diff
@@ -394,11 +391,8 @@ public void close() {
       metrics.statsOnDemandScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTimeMs,
           TimeUnit.MILLISECONDS);
       indexSegmentCount++;
-      if (indexSegmentCount == 1) {
-        logger.info("Completed scanning of 1st index segment by on-demand scanning {} " + indexSegment);
-      }
-      if (indexSegmentCount % 10 == 0) {
-        logger.info("Completed scanning batch of 10 index segments by on-demand scanning {} " + indexSegment);
+      if (indexSegmentCount == 1 || indexSegmentCount % 10 == 0) {
+        logger.info("Index segment processing(on-demand scanning) for compaction stats complete for {}" + indexSegment);
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: remove the "+"

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 9]
```diff
@@ -797,14 +791,9 @@ public void run() {
...
-              if (segmentCount == 1) {
+              if (segmentCount == 1 || segmentCount % 10 == 0) {
                 logger.info(
-                    "Completed scanning 1st sealed segment from last by IndexScanner {} " + indexSegment.getFile()
-                        .getName());
-              }
-              if (segmentCount % 10 == 0) {
-                logger.info("Completed scanning batch of 10 index segments by IndexScanner {} " + indexSegment.getFile()
-                    .getName());
+                    "Completed scanning of sealed segment {} by IndexScanner {} " + indexSegment.getFile().getName());
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: remove the "+"

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 10]
```diff
@@ -775,7 +769,7 @@ public void run() {
         if (!cancelled && indexSegments.size() > 0) {
           Map.Entry<Offset, IndexSegment> activeIndexSegmentEntry = indexSegments.floorEntry(firstCheckpoint);
           long indexSegmentStartProcessTime = time.milliseconds();
-          logger.trace("Fetching active index entries {} in active segments before first checkpoint ",
+          logger.trace("Processing index entries {} in active segments before first checkpoint {}",
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: there are two {} in this log message but only one argument.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java (hunk 9 above)
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: two {} but one argument

**Comment by nsivabalan** [comment]
@vgkholla : you are good to review

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 11]
```diff
@@ -355,6 +357,10 @@ public void close() {
       }
       metrics.statsOnDemandScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTimeMs,
           TimeUnit.MILLISECONDS);
+      indexSegmentCount++;
+      if (indexSegmentCount == 1 || indexSegmentCount % 10 == 0) {
+        logger.info("Index segment processing(on-demand scanning) for container stats complete for {} " + indexSegment);
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: there is a "+" here that needs to be removed. Please take a look and verify all statements.

Also, a space is required b/w "processing" and the opening parentheses

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java [hunk 12]
```diff
@@ -382,6 +390,10 @@ public void close() {
       }
       metrics.statsOnDemandScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTimeMs,
           TimeUnit.MILLISECONDS);
+      indexSegmentCount++;
+      if (indexSegmentCount == 1 || indexSegmentCount % 10 == 0) {
+        logger.info("Index segment processing(on-demand scanning) for compaction stats complete for {}", indexSegment);
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java]: minor: space b/w "processing" and "("

File: ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java [hunk 13]
```diff
@@ -62,6 +64,9 @@ public CompactionDetails getCompactionDetails(long totalCapacity, long usedCapac
         if (bestCandidateToCompact != null) {
           details = new CompactionDetails(validDataSizeByLogSegment.getFirst(),
               bestCandidateToCompact.getSegmentsToCompact());
+          logger.info("Best candidate to compact {} ", bestCandidateToCompact);
+        } else {
+          logger.trace("No best candidate found ");
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java]: minor: sorry to be picky, but there are lot of log messages that have a trailing space at the end (like this one). Could you please remove?

**Pull request merged**